File: backend/PetAdoptionAPI/Models/AdoptionRequest.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;


namespace PetAdoptionAPI.Models
{
    [Table("adoption_requests")]
    public class AdoptionRequest
    {
        [Key]
        [Column("adoption_request_id")]
        public Guid AdoptionRequestId { get; set; }  // Unique ID for the request

        [Column("request_date")]
        public DateTime RequestDate { get; set; } = DateTime.UtcNow;  // When the request was made

        [Column("request_update_date")]
        public DateTime RequestUpdateDate { get; set; } = DateTime.UtcNow;  // When last updated

        [Required]
        [Column("request_status")]
        [StringLength(20)]
        [RegularExpression("^(pending|approved|rejected)$", ErrorMessage = "Invalid request status.")]
        public string RequestStatus { get; set; }  // Status of the request

        [Column("request_message")]
        public string? RequestMessage { get; set; }  // Message from the user (optional)

        [Column("request_response")]
        public string? RequestResponse { get; set; }  // Response from shelter/admin (optional)

        [Column("request_response_date")]
        public DateTime? RequestResponseDate { get; set; }  // When the response was given (optional)

        [Column("request_is_verified")]
        public bool RequestIsVerified { get; set; } = false;  // Whether the request is verified

        [Column("request_is_completed")]
        public bool RequestIsCompleted { get; set; } = false;  // Whether the request is completed

        [Required]
        [Column("user_id")]
        public Guid UserId { get; set; }  // User who made the request

        [Required]
        [Column("animal_id")]
        public Guid AnimalId { get; set; }  // Animal for which the request was made

        // Navigation properties
        [ForeignKey("UserId")]
        public User User { get; set; }

        [ForeignKey("AnimalId")]
        public Animal Animal { get; set; }
    }
}

File: frontend/lib/models/adoption_request.dart
class AdoptionRequest {
  final String adoptionRequestId;
  final DateTime requestDate;
  final DateTime requestUpdateDate;
  final String requestStatus;
  final String? requestMessage;
  final String? requestResponse;
  final DateTime? requestResponseDate;
  final bool requestIsVerified;
  final bool requestIsCompleted;
  final String userId;
  final String animalId;
  final String? animalName; // This will be added later

  // Main constructor
  AdoptionRequest({
    required this.adoptionRequestId,
    required this.requestDate,
    required this.requestUpdateDate,
    required this.requestStatus,
    this.requestMessage,
    this.requestResponse,
    this.requestResponseDate,
    required this.requestIsVerified,
    required this.requestIsCompleted,
    required this.userId,
    required this.animalId,
    this.animalName,
  });

  // Clone with a new animalName (for async fetching)
  AdoptionRequest copyWith({String? animalName}) {
    return AdoptionRequest(
      adoptionRequestId: this.adoptionRequestId,
      requestDate: this.requestDate,
      requestUpdateDate: this.requestUpdateDate,
      requestStatus: this.requestStatus,
      requestMessage: this.requestMessage,
      requestResponse: this.requestResponse,
      requestResponseDate: this.requestResponseDate,
      requestIsVerified: this.requestIsVerified,
      requestIsCompleted: this.requestIsCompleted,
      userId: this.userId,
      animalId: this.animalId,
      animalName: animalName ?? this.animalName,
    );
  }

  // Factory for parsing from JSON
  factory AdoptionRequest.fromJson(Map<String, dynamic> json) {
    return AdoptionRequest(
      adoptionRequestId: json['adoptionRequestId'],
      requestDate: DateTime.parse(json['requestDate']),
      requestUpdateDate: DateTime.parse(json['requestUpdateDate']),
      requestStatus: json['requestStatus'],
      requestMessage: json['requestMessage'],
      requestResponse: json['requestResponse'],
      requestResponseDate: json['requestResponseDate'] != null
          ? DateTime.parse(json['requestResponseDate'])
          : null,
      requestIsVerified: json['requestIsVerified'],
      requestIsCompleted: json['requestIsCompleted'],
      userId: json['userId'],
      animalId: json['animalId'],
      animalName: null, // Will be set later with copyWith
    );
  }
}
